Movie quiz.sprint 6 api

Giya077 opened the pull request.

Giya077 committed:
api: NetworkClient, moviesLoader.
изменил QuestionFactory/Protocol/Delegate
добавил модель: MostPopularMovies

Giya077 committed:
private func showNetworkError доработал при запуске с проблемой сети
    resizedImageURL +

helecta commented on MovieQuiz/Models/MostPopularMovies.swift:
```diff
@@ -0,0 +1,38 @@
+import Foundation
+
+struct MostPopularMovies: Codable {
+    let errorMessage: String
```
Обращу внимание на свойство `errorMessage`. 

Здесь может прийти текст с ошибкой, в таком случае сервис обычно возвращает пустой список фильмов `items`. 
С пустым списком приложение будет выглядеть сломанным. 
Это состояние называют краевым случаем, его стоит обрабатывать как ошибку, например, показывать алерт.
С точки зрения проектирования API сервиса IMDB, 
эту ошибку разработчики должны были вернуть c кодом ошибки клиента ([группа 4хх](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses)). 
Но случается и так, что ответ имеет статус код 200, а тело ответа содержит сообщение об ошибке, как в нашем примере. 
Такой ответ в любом случае необходимо обработать разработчику фронтенда (web, iOS, android).

 **Примеры состояний, когда такое происходит:**

- API key неверный
- API key просрочен
- количество запросов в день превышено (по документации не больше 100)

Giya077 replied:
ок, обработал 3 состояния, сделал отдельный файл ErrorManager, добавил его свойства в метод  didFailToLoadData в основном контролере. Я правильно понял рекомендацию?
<img width="1606" alt="Снимок экрана 2024-01-02 в 14 20 35" src="https://github.com/Giya077/MovieQuiz-ios/assets/150522975/2caf9779-4829-4ec2-8115-2ceaf8568939">
<img width="1335" alt="Снимок экрана 2024-01-02 в 14 21 47" src="https://github.com/Giya077/MovieQuiz-ios/assets/150522975/63fc3db9-cf93-4a9b-87a5-600bbeffcbf7">

helecta commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -8,51 +8,93 @@
+            do {
+                imageData = try Data(contentsOf: movie.resizedImageURL)
+            } catch {
+                print("Failed to load image")
```
Обращу внимание, что игра состоит из вопроса о рейтинге, на который отвечает пользователь, и контекста о фильме в виде постера(картинка). 
Без постера, пользователю будет затруднительно ответить на вопрос о рейтинге фильма верно.

В этом месте, функция не завершает свою работу и идет формировать вопрос без постера. 
В таком случае, бизнес-логика приложения будет нарушена, потому что пользователь не сможет ответить корректно, ведь изображения нет.
Это состояние стоит обработать, например, с помощью алерта. 
Не забудь, что алерт будет обновлять UI и это тоже нужно учесть (вернуть на main), ровно также, как и прерывание функции (сделать return)

Giya077 replied:
как я понял при запросе на следующий вопрос в случае задержки или обрыва сети постер не передается, то должен выходить Алерт с ошибкой,
 и функционал запроса следующего вопроса заблокан до восстановления сети, все сделал. Но при восстановлении сети и получение нового ответа на вопрос квиз начинается заново.

Giya077 replied:
![Uploading Снимок экрана 2024-01-02 в 15.36.23.png…]()

Giya077 replied:
![Uploading Снимок экрана 2024-01-02 в 15.36.23.png…]()

Giya077 replied:
![Uploading qwe.png…]()

Giya077 replied:
<img width="1613" alt="qwe" src="https://github.com/Giya077/MovieQuiz-ios/assets/150522975/4309e582-e608-4f55-b3f5-ecb4fa7e2c6d">

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -144,20 +147,58 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
     }
+    
+    private func showLoadingIndicator() {
+        activityIndicator.isHidden = false // говорим, что индикатор загрузки не скрыт
```
Рекомендую посмотреть на свойство [hidesWhenStopped](https://developer.apple.com/documentation/uikit/uiactivityindicatorview/1622837-hideswhenstopped) 
у `activityIndicator`, и использовать его вместо ручного управления свойством `isHidden`. 

Когда ты будешь вызывать `startAnimating` – индикатор будет появляться автоматически, соответственно на `stopAnimating` – скрываться, а свойство `isHidden` больше не понадобится, индикатор все сделает за тебя. 

**Важно:** это свойство настраивают всего раз раз: кодом во `viewDidLoad()` или аналогичной галочкой у элемента на сториборде.
А строки `activityIndicator.isHidden = false(true)` нужно будет убрать

Giya077 replied:
<img width="1098" alt="Снимок экрана 2024-01-02 в 17 16 56" src="https://github.com/Giya077/MovieQuiz-ios/assets/150522975/64543f05-7f97-4b52-921b-03f664732c8c">

Giya077 replied:
да но я по итогу запутался как надо, в учебнике одно а в ревью по другому

helecta replied:
В учебнике это один из вариантов, я рассказала про другой. И тот и тот вариант правильный.

Извини, я допустила ошибку, почему-то не прожалась кнопка принять твою работу. Увидела я это только сегодня. С твоей работой все хорошо, она отличная. Эти все замечания были необязательными к исправлению, это рекомендации.

Успехов в следущих спринтах!
С Новым годом!

Giya077 replied:
спасибо и тебя так же )
я переделал все

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -144,20 +147,58 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
+    
+    func didLoadDataFromServer() {
+        activityIndicator.isHidden = true // скрываем индикатор загрузки
+        questionFactory?.requestNextQuestion()
```
Ты можешь дополнительно показывать `activityIndicator` между загрузкой вопросов, потому что загрузка вопроса тоже занимает некоторое время из-за получения изображения.
Перед запросом следующего вопроса – показывать индикатор, а в момент получения – выключать анимацию.

Giya077 committed:
обработал 3 состояния, сделал отдельный файл ErrorManager, добавил его свойства в метод didFailToLoadData в основном контролере.
при запросе теперь на следующий вопрос в случае задержки или обрыва сети постер не передается и  выходить Алерт с ошибкой.
